pkg/rid/application/isa.go

```diff
@@ -46,103 +45,112 @@ func (a *app) SearchISAs(ctx context.Context, cells s2.CellUnion, earliest *time
-		return nil, nil, dsserr.PermissionDenied(fmt.Sprintf("ISA is owned by %s", old.Owner))
-	}
-
 	// Validate and perhaps correct StartTime and EndTime.
-	if err := isa.AdjustTimeRange(a.clock.Now(), old); err != nil {
+	if err := isa.AdjustTimeRange(a.clock.Now(), nil); err != nil {
 		return nil, nil, err
 	}
-	// TODO(steeling) do this in a txn.
 	// Update the notification index for both cells removed and added.
 	var (
 		ret  *ridmodels.IdentificationServiceArea
 		subs []*ridmodels.Subscription
 	)
 	// The following will automatically retry TXN retry errors.
-	err = a.Transactor.InTxnRetrier(ctx, func(repo repos.Repository) error {
-		var err error
-		cells := isa.Cells
+	err := a.Transactor.InTxnRetrier(ctx, func(repo repos.Repository) error {
+		// ensure it doesn't exist yet
+		old, err := repo.GetISA(ctx, isa.ID)
+		if err != nil {
+			return err
+		}
 		if old != nil {
-			// TODO steeling, we should change this to a Custom type, to obfuscate
-			// some of these metrics and prevent us from doing the wrong thing.
-			cells = s2.CellUnionFromUnion(old.Cells, isa.Cells)
-			geo.Levelify(&cells)
+			return dsserr.AlreadyExists(fmt.Sprintf("isa with id: %s already exists", isa.ID))
 		}
+
 		// UpdateNotificationIdxsInCells is done in a Txn along with insert since
 		// they are both modifying the db. Insert a susbcription alone does
 		// not do this, so that does not need to use a txn (in subscription.go).
-		subs, err = a.Transactor.UpdateNotificationIdxsInCells(ctx, cells)
-		if err != nil {
-			return err
-		}
-		ret, err = a.Transactor.InsertISA(ctx, isa)
+		subs, err = repo.UpdateNotificationIdxsInCells(ctx, isa.Cells)
 		if err != nil {
 			return err
 		}
-		return nil
+		ret, err = repo.InsertISA(ctx, isa)
+		return err
 	})
 	return ret, subs, err
 }
 
-// UpdateISA implments the AppInterface InsertISA method
+// UpdateISA implments the AppInterface UpdateISA method
 func (a *app) UpdateISA(ctx context.Context, isa *ridmodels.IdentificationServiceArea) (*ridmodels.IdentificationServiceArea, []*ridmodels.Subscription, error) {
-	return nil, nil, dsserr.Internal("not yet implemented")
+	// Update the notification index for both cells removed and added.
+	var (
+		ret  *ridmodels.IdentificationServiceArea
+		subs []*ridmodels.Subscription
+	)
+	// The following will automatically retry TXN retry errors.
+	err := a.Transactor.InTxnRetrier(ctx, func(repo repos.Repository) error {
+		var err error
+
+		old, err := repo.GetISA(ctx, isa.ID)
+		switch {
+		case err != nil:
+			return err
+		case old == nil:
+			return dsserr.NotFound(fmt.Sprintf("isa not found: %s", isa.ID))
+		case old.Owner != isa.Owner:
+			return dsserr.PermissionDenied(fmt.Sprintf("ISA is owned by %s", old.Owner))
+		case !old.Version.Matches(isa.Version):
+			return dsserr.VersionMismatch(fmt.Sprintf("old version for isa: %s", isa.ID))
+		}
+		// Validate and perhaps correct StartTime and EndTime.
+		if err := isa.AdjustTimeRange(a.clock.Now(), old); err != nil {
```

Comment:
nit: We do this outside the transaction for InsertISA -- it would be nice to be consistent here, plus it's less stuff to do inside the transaction.

Reply:
since this uses the old value in adjusttimerange, it needs to be done in the txn, unless we do the version check differently.

In another PR I think we can fix up the logic in adjusttimerange to do what you're saying though